### PR TITLE
Destructuring mutators should handle extra arguments

### PR DESCRIPTION
**Pull request by riknoll**
Deals with a bug where if you used the destructuring mutator (like in the radio packet) and you include extra arguments besides the object being destructured, the mutator would fail to load.

### CODE REVIEW

File: pxtblocks/blocklymutators.ts [hunk 1]
```diff
@@ -66,9 +66,25 @@ namespace pxt.blocks {
 
         switch (mutationType) {
             case MutatorTypes.ObjectDestructuringMutator:
-                if (!info.parameters || info.parameters.length !== 1 || info.parameters[0].properties.length === 0) {
-                    console.error("Mutating blocks only supported for functions with one parameter that has multiple properties");
-                    return;
+                if (!info.parameters || info.parameters.length < 1) {
+                    console.error("Destructuring mutations require at least one parameter")
```
Comment (pelikhan) [inline pxtblocks/blocklymutators.ts]: PXT error

Comment (pelikhan) [inline pxtblocks/blocklymutators.ts]: Localize error message

File: pxtblocks/blocklymutators.ts [hunk 2]
```diff
@@ -66,9 +66,25 @@ namespace pxt.blocks {
...
+                }
+                else {
+                    let found = false;
+                    for (const param of info.parameters) {
+                        if (param.type.indexOf("=>") !== -1) {
+                            if (!param.properties || param.properties.length === 0) {
+                                console.error("Destructuring mutations only supported for functions with an event parameter that has multiple properties");
```
Comment (pelikhan) [inline pxtblocks/blocklymutators.ts]: PXT.error, loc message

File: pxtblocks/blocklymutators.ts [hunk 3]
```diff
@@ -66,9 +66,25 @@ namespace pxt.blocks {
...
+                                return;
+                            }
+                            found = true;
```
Comment (pelikhan) [inline pxtblocks/blocklymutators.ts]: break?

File: pxtblocks/blocklymutators.ts (hunk 1 above)
Comment (riknoll) [inline pxtblocks/blocklymutators.ts]: These are dev errors, not user errors. Only for package authors.

**Comment by riknoll** [comment]
I can add a decompiler test case at least...

**Comment by riknoll** [comment]
@pelikhan  expanding this PR to also include a new feature. This lets APIs set a flag so that when compiling to TS the destructured properties are also passed to the API in an array of enum values. For example, if I chose `arg1` and `arg2`, it would compile to something like this:

```typescript
ns.someCallback([MyArgs.arg1, MyArgs.arg2], ({arg1, arg2}) => {

})
```

That way, blocks can change runtime behavior based on the arguments selected by the user. This involves some complex decompile checking which I have not added a test case for because it is being blocked by an unrelated compiler bug. I have tested it locally and it works fine.

**Comment by pelikhan** [comment]
Could you add docs about that new feature?

**Comment by pelikhan** [comment]
LGTM

**Comment by riknoll** [comment]
Adding docs now

**Comment by pelikhan** [comment]
Do you need to cherry pick into v0?

**Comment by riknoll** [comment]
No cherry-picking required

**Pull request merged**